crates/task-impls/src/transactions.rs

```diff
@@ -243,175 +244,140 @@ impl<
         let task_start_time = Instant::now();
 
         // Find commitment to the block we want to build upon
-        let parent_commitment = self.latest_known_vid_commitment().await;
-
-        let mut latest_block: Option<BuilderResponses<TYPES>> = None;
-        let mut first_iteration = true;
-        while task_start_time.elapsed() < self.api.propose_max_round_time()
-            && latest_block.as_ref().map_or(true, |builder_response| {
-                builder_response
-                    .block_data
-                    .block_payload
-                    .num_transactions(&builder_response.block_data.metadata)
-                    < self.api.min_transactions()
-            })
-        {
-            // Sleep if this isn't the first iteration
-            if first_iteration {
-                first_iteration = false;
-            } else {
-                async_sleep(Duration::from_millis(100)).await;
+        let parent_comm = self.latest_known_vid_commitment().await;
+        let parent_comm_sig = match <<TYPES as NodeType>::SignatureKey as SignatureKey>::sign(
+            &self.private_key,
+            parent_comm.as_ref(),
+        ) {
+            Ok(sig) => sig,
+            Err(err) => {
+                error!(%err, "Failed to sign block hash");
+                return None;
             }
+        };
 
-            let Ok(request_signature) = <<TYPES as NodeType>::SignatureKey as SignatureKey>::sign(
-                &self.private_key,
-                parent_commitment.as_ref(),
-            ) else {
-                error!("Failed to sign block hash");
-                continue;
-            };
-
-            let mut available_blocks = match async_compatibility_layer::art::async_timeout(
+        while task_start_time.elapsed() < self.api.builder_timeout() {
+            match async_compatibility_layer::art::async_timeout(
                 self.api
-                    .propose_max_round_time()
+                    .builder_timeout()
                     .saturating_sub(task_start_time.elapsed()),
-                self.builder_client.get_available_blocks(
-                    parent_commitment,
-                    self.public_key.clone(),
-                    &request_signature,
-                ),
+                self.get_block_from_builder(parent_comm, &parent_comm_sig),
```

Comment:
Ohhhh... so, okay, this doesn't outright make things worse, but... it doesn't make them better either.

So, the implementation here basically took what the builder API was doing, said, "eh, let's ignore all of this and why it was designed this way, and take the most reductive and least efficient approach we can manage". And now that's been tied up with a bow into a nice opaque function.

We have three API calls.

The first should allow HotShot to query any number of builders for their available blocks, and (assuming we ever factor size) wait some amount of time to collect all results that arrive, and choose between the offerings based on fee and size.

The second should fetch one of those blocks - with everything needed to start the next step, e.g. sending it to the DA committee. It should return as close to immediately as it takes to send the response.

The third won't return immediately, not if it's called too fast, because it only starts computing the VID precompile after it builds the block. (note: we may even change it to start computing after the second API request is received)

What this did after the loop wasn't great, but at least it didn't bury the discreet parts even one level of coupling deeper...

Reply:
I'm not sure the "burying one level deeper" is really a problem. The only reason for having a separate function there is to make the retry loop a little simpler. We would want something along these lines regardless, even if it eventually only does the first API call in the inner loop. Other than that, the logic here is exactly the same as before except we won't call the same builder multiple times until we get a certain number of transactions or whatever.

I totally understand what you're saying about the correct way to use the API, and that would absolutely be an improvement. But, that's a bigger change to consensus than I want to make right now (especially the starting DA before VID part) and I'm inclined to go ahead with something along these lines anyways, because
* it will eliminate a variable when debugging the builder (problems arising due to repeat requests, plus this has slightly better timeout handling)
* it will simplify the config down in the sequencer by removing some parameters